Add sandbox checkbox for courses #542

wynnset opened the pull request:
Closes #542

wynnset commented:
This is ready for review @lenglund

andrew-gardener commented on alembic/versions/82ed8fc51745_add_course_sandbox_flag.py:
```diff
@@ -0,0 +1,21 @@
+
+
+def upgrade():
+    op.add_column('course', sa.Column('sandbox', sa.SmallInteger(), nullable=False, server_default='0', default=0));
```
its saver use `with op.batch_alter_table('course', naming_convention=convention) as batch_op:`
for both upgrades and downgrades (especially downgrades since sqlite frequently doesn't like when you drop columns without doing a batch alter in the backend tests)

andrew-gardener replied:
Also please add an index to the column (since it will be used for filtering)

andrew-gardener commented on compair/api/users.py:
```diff
@@ -377,6 +379,12 @@ def get(self):
                         Course.term.like(search)
                     ))
+
+        if params['includeSandbox'] == "false":
```
can you make this flexible so like so:
```
if params['sandbox']:
    query = query.filter(Course.sandbox=params['sandbox'])
```

This allow sandbox to not effect course listing by default and also allow you to display only sandbox courses and non-sandbox courses

andrew-gardener commented on compair/api/users.py:
```diff
@@ -422,6 +430,11 @@ def get(self, user_uuid):
                         Course.term.like(search)
                     ))
 
+        if params['includeSandbox'] == "false":
```
same as other `includeSandbox`

andrew-gardener commented on compair/models/course.py:
```diff
@@ -18,6 +18,7 @@ class Course(DefaultTableMixin, UUIDMixin, ActiveMixin, WriteTrackingMixin):
     name = db.Column(db.String(255), nullable=False)
     year = db.Column(db.Integer, nullable=False)
     term = db.Column(db.String(255), nullable=False)
+    sandbox = db.Column(db.SmallInteger(), nullable=False, server_default='0', default=0)
```
you can use`db.Boolean(name='sandbox')` (I think it works a bit better across different databases and also uses `TINYINT` for mysql). Would need to upgrade the migration as well

andrew-gardener commented on compair/static/modules/course/course-assignments-partial.html:
```diff
@@ -3,6 +3,7 @@
     <div class="row">
         <header class="col-sm-6">
             <h1><i class="fa fa-book"></i> {{course.name}} <br />({{course.year}} {{course.term}})</h1>
+            <span ng-show="course.sandbox" class="label label-default">Sandbox Course</span>
```
@lenglund should the name be overridden if sandbox?

andrew-gardener commented on compair/static/modules/assignment/assignment-module_spec.js:
```diff
@@ -62,6 +62,7 @@ describe('assignment-module', function () {
         "name": "Test Course",
         "year": 2015,
         "term": "Winter",
+        "sandbox": 0,
```
please use `false` instead of `0` for the specs (helps to keep track of what type things are

andrew-gardener commented on compair/static/modules/home/home-partial.html:
```diff
@@ -20,10 +20,16 @@
                 <input class="form-control" type="text" name="search" placeholder="Filter courses by keyword"
                        ng-model="courseFilters.search">
+                <br>
+                <label for="includeSandbox">
```
Can you use a dropdown instead with default `All Courses` (`courseFilters.sandbox = false`) and second option `All Sandbox Courses`  (`courseFilters.sandbox = true`)

andrew-gardener commented on compair/static/modules/user/user-course-partial.html:
```diff
@@ -3,10 +3,16 @@
                 <input class="form-control" type="text" name="search" placeholder="Filter courses by keyword"
                        ng-model="courseFilters.search">
+                <br>
+                <label for="includeSandbox">
```
same as homepage change

andrew-gardener commented on compair/api/users.py:
```diff
@@ -61,9 +61,11 @@ def non_blank_text(value):
 
 user_course_list_parser = pagination_parser.copy()
 user_course_list_parser.add_argument('search', required=False, default=None)
+user_course_list_parser.add_argument('includeSandbox', required=False, default=None)
```
add `type=bool`

andrew-gardener commented on compair/api/users.py:
```diff
@@ -61,9 +61,11 @@ def non_blank_text(value):
 
 user_id_course_list_parser = pagination_parser.copy()
 user_id_course_list_parser.add_argument('search', required=False, default=None)
+user_id_course_list_parser.add_argument('includeSandbox', required=False, default=None)
```
add `type=bool`

andrew-gardener commented on compair/api/users.py:
```diff
@@ -377,6 +379,12 @@ def get(self):
                         Course.term.like(search)
                     ))
+
+        if params['includeSandbox']:
```
change to `if params['includeSandbox'] != None:`

andrew-gardener commented on compair/api/users.py:
```diff
@@ -377,6 +379,12 @@ def get(self):
+
+        if params['includeSandbox']:
+            query = query.filter(
+                Course.sandbox == (params['includeSandbox'] != "false" and params['includeSandbox'] != "0")
```
when the type is bool you can use `Course.sandbox == params['includeSandbox']`

andrew-gardener commented on compair/api/users.py:
```diff
@@ -422,6 +430,11 @@ def get(self, user_uuid):
                         Course.term.like(search)
                     ))
 
+        if params['includeSandbox']:
```
update this one too

andrew-gardener commented on compair/models/course.py:
```diff
@@ -18,6 +18,7 @@ class Course(DefaultTableMixin, UUIDMixin, ActiveMixin, WriteTrackingMixin):
     name = db.Column(db.String(255), nullable=False)
     year = db.Column(db.Integer, nullable=False)
     term = db.Column(db.String(255), nullable=False)
+    sandbox = db.Column(db.Boolean(name='sandbox'), nullable=False)
```
add default value here to get tests to pass (recommend `False`)

andrew-gardener commented on alembic/versions/82ed8fc51745_add_course_sandbox_flag.py:
```diff
@@ -0,0 +1,23 @@
+
+def downgrade():
+    with op.batch_alter_table('course', naming_convention=convention) as batch_op:
+        batch_op.drop_column('sandbox');
```
add in `batch_op.drop_index('ix_course_sandbox')` to get tests to pass

andrew-gardener replied:
nit also remove semi-colon

andrew-gardener commented on alembic/versions/82ed8fc51745_add_course_sandbox_flag.py:
```diff
@@ -0,0 +1,23 @@
+from compair.models import convention
+
+def upgrade():
+    op.add_column('course', sa.Column('sandbox', sa.Boolean(name='sandbox'), nullable=False, server_default='0', default=False, index=True));
```
nit remove semi-colon

andrew-gardener commented on compair/models/course.py:
```diff
@@ -18,6 +18,7 @@ class Course(DefaultTableMixin, UUIDMixin, ActiveMixin, WriteTrackingMixin):
     name = db.Column(db.String(255), nullable=False)
     year = db.Column(db.Integer, nullable=False)
     term = db.Column(db.String(255), nullable=False)
+    sandbox = db.Column(db.Boolean(name='sandbox'), nullable=False, default=False)
```
add `index=True` here as well

wynnset commented:
@lenglund I noticed you have removed the ability to filter courses, which is a major part of what I worked on for this ticket. @andrew-gardener had mentioned to me that we wanted the ability to filter the courses, is this not the case?

wynnset commented:
I've added the dropdown on homepage for admins only, and also fixed an issue with the filtering not working properly. I'll merge this in tomorrow. Let me know in the meantime if you have any objections or require further changes.

andrew-gardener commented on compair/static/modules/home/home-partial.html:
```diff
@@ -18,12 +18,30 @@
+                ng-model="courseFilters.search">
+        </form>
+
+        <form class="col-sm-6 search-courses text-right" role="search" ng-show="canManageUsers && canAddCourse && (courses.length || courseFilters.search || courseFilters.includeSandbox!='')" style="padding-right:0;">
```
The conditions on this need to be split. I believe it should be 
`canAddCourse && (courses.length || courseFilters.search || courseFilters.includeSandbox!=null)`

wynnset replied:
@andrew-gardener what's the purpose of splitting this condition?

andrew-gardener replied:
instructors (having `canAddCourse` but not `canManageUsers`) can no longer search

wynnset replied:
There's another form element for instructors just before this form element.

andrew-gardener commented on compair/static/modules/home/home-partial.html:
```diff
@@ -18,12 +18,30 @@
+        </form>
+
+        <form class="col-sm-6 search-courses text-right" role="search" ng-show="canManageUsers && canAddCourse && (courses.length || courseFilters.search || courseFilters.includeSandbox!='')" style="padding-right:0;">
+            <div class="input-group">
```
then add an ng-show with condition `canManageUsers`

andrew-gardener commented on compair/static/modules/home/home-partial.html:
```diff
@@ -122,7 +144,8 @@ <h3 class="media-heading">
     <div ng-if="!courses.length">
         <hr />
-        <p ng-if="!courseFilters.search">No courses currently available. <span ng-if="canAddCourse">Would you like to <a href="#/course/create">add a course</a>?</span></p>
+        <p ng-if="!courseFilters.search && courseFilters.includeSandbox==''">No courses currently available. <span ng-if="canAddCourse">Would you like to <a href="#/course/create">add a course</a>?</span></p>
```
need to update this to `courseFilters.includeSandbox==null`

andrew-gardener commented on compair/static/modules/home/home-partial.html:
```diff
@@ -122,7 +144,8 @@ <h3 class="media-heading">
         <hr />
-        <p ng-if="!courseFilters.search">No courses currently available. <span ng-if="canAddCourse">Would you like to <a href="#/course/create">add a course</a>?</span></p>
+        <p ng-if="!courseFilters.search && courseFilters.includeSandbox==''">No courses currently available. <span ng-if="canAddCourse">Would you like to <a href="#/course/create">add a course</a>?</span></p>
+        <p ng-if="!courseFilters.search && courseFilters.includeSandbox!=''">No courses found. Please try changing to "All courses" above.</p>
```
and this to `courseFilters.includeSandbox!=null`

andrew-gardener commented on compair/static/modules/user/user-course-partial.html:
```diff
@@ -1,12 +1,25 @@
 
-        <form class="col-sm-4 search-courses text-right" role="search" ng-show="courses.length || courseFilters.search">
-            <div class="form-group">
+        <form class="col-sm-6 search-courses text-right" role="search" ng-show="courses.length || courseFilters.search || courseFilters.includeSandbox!=''">
```
`courseFilters.includeSandbox!=null`

andrew-gardener commented on compair/static/modules/user/user-course-partial.html:
```diff
@@ -73,7 +87,8 @@ <h1 class="col-sm-8">Manage {{ user.displayname }}'s Courses</h1>
     <div ng-if="!courses.length">
         <hr />
-        <p ng-if="!courseFilters.search">No courses found. Select <a href="#/users">another user</a>?</p>
+        <p ng-if="!courseFilters.search && courseFilters.includeSandbox==''">No courses found. Select <a href="#/users">another user</a>?</p>
```
`courseFilters.includeSandbox==null`

andrew-gardener commented on compair/static/modules/user/user-course-partial.html:
```diff
@@ -73,7 +87,8 @@ <h1 class="col-sm-8">Manage {{ user.displayname }}'s Courses</h1>
         <hr />
-        <p ng-if="!courseFilters.search">No courses found. Select <a href="#/users">another user</a>?</p>
+        <p ng-if="!courseFilters.search && courseFilters.includeSandbox==''">No courses found. Select <a href="#/users">another user</a>?</p>
+        <p ng-if="!courseFilters.search && courseFilters.includeSandbox!=''">No courses found. Please try another filter and/or search term above.</p>
```
`courseFilters.includeSandbox!=null`
